### ISSUE TITLE
Copying to/from system clipboard doesn't work

### ISSUE THREAD

**Issue by zetok** [neovim/neovim#2889]
OS: Gentoo
WM: `i3-4.10.2`
Xorg: `1.16.4`
Neovim:
```
$ vi --version
NVIM 0.0.0-alpha+201506240043 (compiled Jun 24 2015 10:43:29)
Commit: 91fcd3378750080fb94d3030a6fc29a4f86b15e8
Build type: Gentoo
Compilation: /usr/bin/x86_64-pc-linux-gnu-gcc -O2 -march=amdfam10 -ggdb -pipe -Wno-error -Wconversion  -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -DINCLUDE_GENERATED_DECLARATIONS -DHAVE_CONFIG_H -I/var/tmp/portage/app-editors/neovim-9999/work/neovim-9999_build/config -I/var/tmp/portage/app-editors/neovim-9999/work/neovim-9999/src -I/usr/include -I/usr/include -I/usr/include/luajit-2.0 -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/var/tmp/portage/app-editors/neovim-9999/work/neovim-9999_build/src/nvim/auto -I/var/tmp/portage/app-editors/neovim-9999/work/neovim-9999_build/include
Compiled by portage@galaxor

Optional features included (+) or not (-): +acl   +iconv    -jemalloc 
For differences from Vim, see :help vim-differences

   system vimrc file: "/etc/vim/nvimrc"
     user vimrc file: "~/.nvimrc"
 2nd user vimrc file: "~/.nvim/nvimrc"
      user exrc file: "~/.exrc"
  fall-back for $VIM: "/usr/share/nvim"
```

====

No special clipboard provider is being used.

When trying to paste from clipboard (``"+P``):
```
clipboard: No shell command for communicating with the clipboard found.
clipboard: provider is not available
Press ENTER or type command to continue
```
After pressing ``ENTER``:
```
clipboard: provider is not available
E353: Nothing in register +
Press ENTER or type command to continue
```

~same error when trying to copy to clipboard.

====

Copying to/from system clipboard works in Vim.

Is there any other info could I provide?

**Comment by bfredl** [comment]
You need `xsel` or `xclip` installed, see `:help nvim-clipboard`. Perhaps the error message should be updated to say `see :help nvim-clipboard`.

**Comment by zetok** [comment]
@bfredl if I understood correctly, Neovim is not going to ``just work™ OOTB``, and requires user to do stuff manually in order to even get clipboard working?

Furthermore, it executes shell commands to just get stuff to/from clipboard?

**Comment by Pyrohh** [comment]
> if I understood correctly, Neovim is not going to just work™ OOTB, and requires user to do stuff manually in order to even get clipboard working?

Yes.

> Furthermore, it executes shell commands to just get stuff to/from clipboard?

Yes, that way we don't need to link to xlib, among other things. Unfortunately, `/dev/snarf` never caught on, so I think this is the most reasonable solution.

**Comment by RedX2501** [comment]
What provider do i need to install to get it to work on windows?

**Comment by justinmk** [comment]
> Neovim is not going to just work™ OOTB, and requires user to do stuff manually in order to even get clipboard working?

It works OOTB if you have `xsel` or `xclip` Why would you have X Window but not `xsel` or `xclip`?

> it executes shell commands to just get stuff to/from clipboard?

Yes, to avoid a compile-time dependency on X.

**Comment by ZyX-I** [comment]
@justinmk They are not installed by default by a number of distributions and definitely both are not part of packages that provide X server. From my experience having Python and pip already installed is more likely (+ a number of plugins that require having Python support), so it may be good idea to try using Python ctypes with X library.

**Comment by bfredl** [comment]
> Python ctypes with X library.

If we do that, why not just download/build [xsel-vim](https://github.com/bfredl/xsel/tree/nvim) as part of the buildsystem, conditionally if `libX11` is available (and os is not OS X) ? 
If better performance is needed, I would imagine supporting in-process binary plugins. Such plugin would be trivial (at least, no event-handling) for OS X, and for X11 it would need being able to register a callback for polling on a fd. This would still isolate the dependency as only the loadable plugin would depend on `libX11.so` and not nvim binary. But I could imagine the performance problem being solvable in the surrounding logic: things like `[getreg('*'), getregtype('*')]`  should ideally generate only one provider request.

**Comment by zetok** [comment]
@justinmk
>>Neovim is not going to just work™ OOTB, and requires user to do stuff manually in order to even get clipboard working?

>It works OOTB if you have `xsel` or `xclip` Why would you have X Window but not `xsel` or `xclip`?

Why would I have them if so far I didn't have a need for them? To this day I didn't even know about their existence…

It's not working OOTB, since there's nothing said in FAQ, etc about Neovim having a runtime dependency on either of those to get basic functionality such as clipboard… @bfredl's suggestion to have at least more helpful error message would help, but it still wouldn't be OOTB.


@bfredl 
>> Python ctypes with X library.

>If we do that, why not just download/build [xsel-vim](https://github.com/bfredl/xsel/tree/nvim) as part of the buildsystem, conditionally if `libX11` is available (and os is not OS X) ? 

Um, whatever you're going to do, please keep in mind that there are build systems that sandbox networking during building, so that anything depending on network connection would fail to build.. one of those is portage, which has ``FEATURE="network-sandbox"``

Also, if possible, please don't make building that dependant only on presence of ``libX11``, since there are cases where one might not actually want to have support for xorg stuff in Neovim, even if X is in the system. E.g., I don't know much about Wayland, but I suspect that its compositors don't depend on ``libX11`` to for clipboard, and it's quite likely that someone who's using Wayland has xorg installed by side.

I.e. if possible, a build-time switch for it would be nice to have.

**Comment by bfredl** [comment]
Sorry, when I said `buildsystem` I meant the wrapper Makefile that downloads and build a lot of dependencies (libuv, busted for testing, etc). In a buildscript/specification for distributions I would imagine it being handled as a proper dependency or part of the sources. (I don't really know what the existing buildscripts do for all fetched dependencies)

And you're right wayland makes the picture a bit more complicated.

**Comment by justinmk** [comment]
> please keep in mind that there are build systems that sandbox networking during building, 

Distros can specify a dependency on `xsel` or whatever in their respective package systems. This doesn't require any action from the Neovim project.

> there are cases where one might not actually want to have support for xorg stuff in Neovim, even if X is in the system. E.g., I don't know much about Wayland, but I suspect that its compositors don't depend on libX11 for clipboard, and it's quite likely that someone who's using Wayland has xorg installed by side.

Which is one reason the provider system is nice. We can add providers for Wayland in the future, without changing the core.

**Comment by Pyrohh** [comment]
> Distros can trivially specify a dependency on xsel or whatever in their respective package systems. This doesn't require any action from the Neovim project.

I don't think that was @zetok's point, which was in response to @bfredl mentioning his fork, `xsel-vim`. We could formally create a fork of xsel and get it packaged, but that seems... excessive.

If we're going to include it, then the source should probably just be packaged with nvim (it's only a few files IIRC). Of course it'd be better if @bfredl's PR would just be accepted, but the xsel maintainer seems to be unresponsive regarding it.

**Comment by justinmk** [comment]
Whatever is blocked by https://github.com/kfish/xsel/pull/8 is simply blocked. That's out of our control for now.

**Comment by Pyrohh** [comment]
Closing this as the issue doesn't reflect the title.

**Comment by HarishVangala** [comment]
Just for clarity:

**Problem:** Unable to use system clipboard via + register. Usually people use a command like "+yy which fails with an error message like `clipboard provider is not available neovim` etc

**Solution:** Simply install an application called `xclip` (i.e. run for e.g. `sudo apt-get install xclip`). That's it! Neovim implicitly takes care of the problem. You may have to restart neovim (`nvim`).

**Reasons:**  As taken from `:help clipboard` run from neovim commandline

> Nvim has no direct connection to the system clipboard. Instead it is
> accessible through a |provider| which transparently uses shell commands for
> communicating with the clipboard.
> 
> Clipboard access is implicitly enabled if any of the following clipboard tools
> are found in your `$PATH`.
> 
>   - xclip
>   - xsel (newer alternative to xclip)
>   - pbcopy/pbpaste (Mac OS X)
>   - lemonade (for SSH) https://github.com/pocke/lemonade
>   - doitclient (for SSH) http://www.chiark.greenend.org.uk/~sgtatham/doit/

As on 5 October, 2016 on Ubuntu 16.04LTS, neovim installed via ppa.

**Comment by ivangjxyz** [comment]
If someone is getting here, and they are already using Wayland, the trick is to install `wl-clipboard` to make it access the system clipboard. 

No need for `xclip` or `xsel`.

**Comment by mehadihasanrajib10** [comment]
> You need `xsel` or `xclip` installed, see `:help nvim-clipboard`. Perhaps the error message should be updated to say `see :help nvim-clipboard`.
Thanks Brother This is Work for me .

**Comment by xiyini** [comment]
> If someone is getting here, and they are already using Wayland, the trick is to install `wl-clipboard` to make it access the system clipboard.
> 
> No need for `xclip` or `xsel`.

You're a life safer!
